core/matrix/diagonal.cpp

```diff
@@ -302,4 +302,24 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_TYPE(GKO_DECLARE_DIAGONAL_MATRIX);
 
 
 }  // namespace matrix
+
+
+// Implement DiagonalExtractable for LinOp when Diagonal is complete class
+template <typename ValueType>
+std::unique_ptr<LinOp> DiagonalExtractable<ValueType>::extract_diagonal_linop()
+    const
+{
+    auto diag = this->extract_diagonal();
```

Comment:
Am I missing something here ? `Diagonal` does not have a `extract_diagonal` method, right ?

Reply:
no, but the DiagonalExtractable use `as` or `dynamic_cast` to convert the output from Diagonal to Linop.
then it requires Diagonal be a complete class.
I put the implementation into Diagonal and instantiate it.